nengo_viz/static/viz_netgraph.js

```diff
@@ -0,0 +1,243 @@
+/**
+ * Network diagram
+ * @constructor
+ *
+ * @param {dict} args - A set of constructor arguments, including:
+ * @param {int} args.id - the id of the server-side NetGraph to connect to
+ * @param {DOMElement} args.parent - the element to add this component to
+ */
+VIZ.NetGraph = function(args) {
+
+    /** create the master SVG element */
+    this.svg = this.createSVGElement('svg');
+    this.svg.classList.add('netgraph');    
+    this.svg.style.width = '100%';
+    this.svg.style.height = 'calc(100% - 80px)';
+    this.svg.style.position = 'fixed';
+    args.parent.appendChild(this.svg);
+    this.parent = args.parent;
+    
+    /** three separate layers, so that expanded networks are at the back,
+     *  then connection lines, and then other items (nodes, ensembles, and
+     *  collapsed networks) are drawn on top. */
+    this.g_networks = this.createSVGElement('g'); 
+    this.svg.appendChild(this.g_networks);
+    this.g_conns = this.createSVGElement('g');
+    this.svg.appendChild(this.g_conns);
+    this.g_items = this.createSVGElement('g');
+    this.svg.appendChild(this.g_items);
+    
+    /** connect to server */
+    this.ws = new WebSocket('ws://localhost:8080/viz_component?id=' + args.id);
+    this.ws.binaryType = "arraybuffer";
+    this.ws.onmessage = function(event) {self.on_message(event);}
+
+    /** respond to resize events */
+    this.svg.addEventListener("resize", function() {self.on_resize();});
+    window.addEventListener("resize", function() {self.on_resize();});
+    
+    this.scale = 1.0;          // global scaling factor
+    this.offsetX = 0;          // global x,y pan offset 
+    this.offsetY = 0;
+
+    this.svg_objects = {};     // dict of all VIZ.NetGraphItems, by uid
+    this.svg_conns = {};       // dict of all VIZ.NetGraphConnections, by uid
+
+    /** Since connections may go to items that do not exist yet (since they
+     *  are inside a collapsed network), this dictionary keeps a list of
+     *  connections to be notified when a particular item appears.  The
+     *  key in the dictionary is the uid of the nonexistent item, and the
+     *  value is a list of VIZ.NetGraphConnections that should be notified
+     *  when that item appears. */
+    this.collapsed_conns = {}; 
+    
+    /** dragging the background pans the full area by changing offsetX,Y */
+    var self = this;
```

Comment:
Fair enough -- it'd be a good standard to stick with to try to define variables like that as close to the top of the constructor as possible...  lemme rearrange it...